### PR TITLE
fix(core): bound snapshot deserialization + property tests (#254 wave1)

### LINKED ISSUE #411: [super-qa] medium: Unbounded MessagePack deserialization of .snapshot sidecar (allocation DoS)
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis (security lens ran via octo tier-2 fallback, xhigh not effort:max); **not independently source-verified**. Review before action.

**Unbounded MessagePack deserialization of .snapshot sidecar (allocation DoS)**

- **ID:** `engine/security-snapshot-unbounded-deserialize`
- **Severity:** medium
- **Area:** core · **Type:** security
- **Location:** `src/engine/snapshot.rs:277 (also 241, 210)`
- **Auto-fixable:** no

## Description
`load_from_file()` reads the entire `.snapshot` sidecar into memory with `fs::read(path)` (line 210, no size cap) and then deserializes the payload with `rmp_serde::from_slice` (line 277) into `SnapshotPayload`, which contains `Vec<GraphEdgeSnapshot>`, `Vec<ScopeNode>`, and `Vec<HnswEntry>` (each `HnswEntry` holding a `Vec<f32>` embedding). rmp-serde 1.3.1 honors the MessagePack array-length prefix via serde's `size_hint`, so the derived `Visitor` performs `Vec::with_capacity(declared_len)` BEFORE consuming elements. A crafted file that declares a huge array length forces a large up-front allocation (memory-exhaustion / OOM-abort DoS) even though the file itself is small.

The blake3 checksum (verified at line 270, before payload deserialize at 277) is the only gate, but blake3 here is UNKEYED and serves as a corruption/integrity check, not an authentication control: any actor able to write the sidecar can recompute a matching blake3 over a malicious payload. The header is deserialized at line 241 BEFORE the checksum check, but `SnapshotHeader` holds only scalar fields, so the amplification risk is confined to the payload path.

Threat model: exploitation requires write access to the `<db>.snapshot` file in the engine's own data directory, so this is a local tamper / shared-data-dir DoS rather than a remote one — hence medium, not high. The flat (non-recursive) snapshot types mean there is no stack-overflow/recursion-depth angle here; the concern is purely unbounded allocation. CWE-502, CWE-400, CWE-789.

## Suggested approach
Cap both the file size and the per-collection element counts before/while deserializing. Minimal mechanical guard at the file level:

```rust
const MAX_SNAPSHOT_BYTES: u64 = 512 * 1024 * 1024; // tune to expected corpus
let meta = fs::metadata(path).ok()?;
if meta.len() > MAX_SNAPSHOT_BYTES {
    tracing::warn!(path = %path.display(), size = meta.len(), "snapshot exceeds size cap, discarding");
    return None;
}
let data = fs::read(path).ok()?; // already capped
```

The robust fix additionally bounds the in-band declared lengths so a small file cannot pre-allocate a huge Vec: deserialize through a length-limited reader/visitor, or post-validate `payload.graph.edges.len()`, `payload.scope_tree.nodes.len()`, and each `HnswEntry::embedding.len() == embed_dim` (the embedding-dim check already exists in ann.rs:191 but only on the ann load path). Because blake3 is unkeyed it cannot be relied on as an authenticity gate; if sidecar authenticity matters, use a keyed MAC (blake3::keyed_hash) with a key not stored beside the file.

## References
CWE-502, CWE-400, CWE-789

### LINKED ISSUE #450: [super-qa] medium: Pure temporal helpers `passes_temporal_cutoff` and `fact_overlaps_period` lack property-based tests
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**Pure temporal helpers `passes_temporal_cutoff` and `fact_overlaps_period` lack property-based tests**

- **ID:** `engine/proptest-temporal-helpers`
- **Severity:** medium
- **Area:** core · **Type:** test
- **Location:** `src/engine/mod.rs:690-708`
- **Auto-fixable:** no

## Description
`passes_temporal_cutoff` (line 690) and `fact_overlaps_period` (line 705) are free functions with clear algebraic invariants — interval containment and overlap — but have only implicit coverage from example-based query tests. No proptest verifies: (a) `passes_temporal_cutoff` is monotone in `cutoff` (if it passes at T it passes at T+ε when `t_invalid` is unbounded); (b) `fact_overlaps_period(f, s, e)` ↔ `fact_overlaps_period(f, s', e')` for all sub-intervals; (c) a fact with `t_valid > cutoff` never passes. These are the kind of off-by-one errors (< vs <=) that example tests routinely miss.

## Suggested approach
In `src/engine/mod.rs` (or a companion `tests/proptest_temporal.rs`), add:
```rust
#[cfg(test)]
mod proptest_temporal {
    use proptest::prelude::*;
    use chrono::{DateTime, TimeZone, Utc};
    // Strategy: generate timestamps as i64 seconds
    prop_compose! {
        fn arb_ts()(s in 0i64..=4_000_000_000) -> DateTime<Utc> {
            Utc.timestamp_opt(s, 0).unwrap()
        }
    }
    proptest! {
        #[test]
        fn cutoff_monotone(t_valid in proptest::option::of(arb_ts()),
                           cutoff in arb_ts(),
                           delta in 1i64..=1000) {
            // If fact passes at cutoff, it must pass at cutoff + delta
            // when t_invalid is None
            let fact = make_fact(t_valid, None);
            if passes_temporal_cutoff(&fact, cutoff) {
                prop_assert!(passes_temporal_cutoff(&fact, Utc.timestamp_opt(cutoff.timestamp() + delta, 0).unwrap()));
            }
        }
    }
}
```

## References
CWE-682

### LINKED ISSUE #451: [super-qa] medium: Snapshot serde roundtrip covered by a single example test; no property-based roundtrip
> ⚠️ Generated by `/super-qa` 2026-06-01. Severity & details from agent analysis; **not independently source-verified**. Review before action.

**Snapshot serde roundtrip covered by a single example test; no property-based roundtrip**

- **ID:** `engine/proptest-snapshot-roundtrip`
- **Severity:** medium
- **Area:** core · **Type:** test
- **Location:** `src/engine/snapshot.rs`
- **Auto-fixable:** no

## Description
`snapshot.rs` has a single example-based roundtrip test (the happy path with one fact). The `write_to_file` → `load_from_file` pipeline involves two msgpack serializations, blake3 integrity check, and a u32 LE header length field. A proptest over arbitrary `SnapshotHeader`/`SnapshotPayload` fields (dimension, fact count, timestamps at boundary values like `i64::MIN`, `i64::MAX`) would catch endianness and overflow bugs that example tests miss. The current coverage also leaves the following `load_from_file` branches untested by proptest: truncated file, corrupted hash, mismatched embed_dim.

## Suggested approach
Add a proptest in `snapshot.rs`'s `#[cfg(test)]` block that generates random valid fact counts (0..=100), write to a tempfile, and assert `load_from_file` returns `Some` with identical data. Separately, add a proptest that bit-flips a random byte and asserts `load_from_file` returns `None` (blake3 mismatch).

## References
—

### PR DESCRIPTION
**Pull request by dutiona**
Part of the **#254** super-qa `area:core` sweep (Wave 1). Subsystem-cohesive PR closing 3 findings.

## Findings closed

- **#411** [security/medium] — Unbounded MessagePack deserialization of .snapshot sidecar (allocation DoS)
- **#451** [test/low] — Snapshot serde roundtrip lacks property-based + bit-flip proptest
- **#450** [test/low] — Temporal helpers passes_temporal_cutoff / fact_overlaps_period lack property-based tests

## Review (internal diverse-lens subagents — no external models)

3 clean-slate adversarial reviewers (correctness / security & soundness / api-surface & test-quality): **0 blocker, 3 major** found.

All blocker/major **addressed** in fix commits (3):
- Finding 1 (major, correctness): Rewrote the MAX_SNAPSHOT_BYTES doc-comment and the load_from_file inline comment in /home/mroynard/dev/memory-engine/.worktrees/sweep-254-snapshot/src/engine/snapshot.rs. Removed the false claim that rmp_serde pre-allocates Vecs from untrusted in-band MessagePack leng
- Finding 2 (major, correctness): file_at_size_cap_is_not_rejected_for_size (renamed size_cap_boundary_is_exclusive) no longer creates a 512 MiB sparse file or calls load_from_file (which would fs::read 512 MiB of zeros into a Vec<u8>, risking OOM under cargo's parallel harness). The exclusive (> not 
- Finding 3 (major, api-surface & test-quality): Extracted the pre-read metadata/size check from load_from_file into a private testable seam size_gate(path, cap) -> SizeGate (Proceed/Reject). New unit test size_gate_rejects_over_cap_before_read exercises the exact over-cap branch with a 4 KiB file and

## Gate
`cargo build/test/clippy --workspace` + `cargo fmt --check` — green (rebased on current main).

Fixes #411, fixes #451, fixes #450

### CODE REVIEW

File: src/engine/snapshot.rs [hunk 1]
```diff
@@ -31,6 +31,74 @@ pub const FORMAT_VERSION: u32 = 1;
...
+        }
+        Ok(_) => SizeGate::Proceed,
+        Err(e) => {
+            if e.kind() != std::io::ErrorKind::NotFound {
+                tracing::warn!(path = %path.display(), error = %e, "snapshot metadata failed");
+            }
+            SizeGate::Reject
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/engine/snapshot.rs:110]: Done in 4f7bf5a — size_gate now rejects non-regular-file paths (is_file check ordered before the size check) with a clear warning, so a directory is rejected up front instead of failing later in fs::read. Added a dir-rejection unit test.
